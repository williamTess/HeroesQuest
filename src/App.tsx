import React, { useEffect, useState } from "react";
import { FileUploader } from "react-drag-drop-files";
import logo from "./logo.svg";
import "./App.css";
import readFile from "./utils/readFile";
import { createMap } from "./utils/createMap";
import RenderMap from "./utils/RenderMap";
import { Hero } from "./type/type";
import { onMove } from "./utils/onMove";
import { RenderResults } from "./utils/RenderResults";

const App = () => {
  const [content, setContent] = useState<string[]>([]);
  const [map, setMap] = useState<string[][]>([]);
  const [heroes, setHeroes] = useState<Hero[]>([]);
  const [error, setError] = useState<string>("");
  const [isDone, setIsDone] = useState<boolean>(false);

  useEffect(() => {
    if (content.length > 0) {
      const { map, heroes, error } = createMap(content);
      setError(error);
      setMap(map);
      setHeroes(heroes);
    }
  }, [content]);

  const onSubmitFile = (file: any) => {
    const reader = readFile(file);
    if (reader) {
      reader.onload = () => {
        const tmp = reader.result as string;
        setContent(tmp.split("\n"));
      };
    } else {
      alert("error read file");
    }
  };

  const onClickStart = () => {
    let tmpMap = [...map];
    const newHeroes: Hero[] = [];
    heroes.forEach((hero) => {
      let currentStateHero = { ...hero };
      hero.moves.forEach((move) => {
        const { map, newHeroStat } = onMove(tmpMap, currentStateHero, move);
        tmpMap = map;
        currentStateHero = { ...newHeroStat };
      });
      newHeroes.push(currentStateHero);
    });
    setHeroes(newHeroes);
    setMap(tmpMap);
    setIsDone(true);
  };

  if (isDone)
    return (
      <div className="App">
        <RenderResults map={map} heroes={heroes} />
      </div>
    );

  return (
    <div className="App">
      {map.length > 0 && !error ? (
        <>
          <RenderMap map={map} />
          <button onClick={onClickStart}>start</button>
        </>
      ) : (
        <>
          <img src={logo} className="App-logo" alt="logo" />

          <FileUploader
            handleChange={onSubmitFile}
            name="file"
            types={["TXT"]}
          />
          {error}
        </>
      )}
    </div>
  );
};

export default App;
