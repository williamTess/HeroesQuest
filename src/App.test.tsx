import React from "react";
import { render, screen } from "@testing-library/react";
import App from "./App";
import ReactDOM from "react-dom/client";
import { root } from ".";
import { shallow } from "enzyme";

test("Open main page", () => {
  const component = shallow(<App />);
  expect(component).toMatchSnapshot();
});

// test("index should render", () => {
//   const div = document.createElement("div");
//   div.id = "root";
//   document.body.appendChild(div);
//   expect(root.render).toHaveBeenCalledWith(<App />, div);
// });
