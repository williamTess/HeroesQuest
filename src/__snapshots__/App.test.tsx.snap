// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Open main page 1`] = `ShallowWrapper {}`;
